<!DOCTYPE html>
<html>
<head>	
	<title>MGRA server</title>
	<style>
		 #menu { 
    		float: left; 
    		width: 15%;
    		padding: 5px; 
		    margin: 10px 10px 20px 5px; 
   		}

		#content_text { 
			float: left; 
		    margin: 2px 5px 0px 5px; 
		    width: 150px; 
   		}

		#content_text1 { 
			float: left; 
		    margin: 10px 5px 20px 5px; 
		    width: 220px; 
   		}

		#content {
		    margin: 10px 5px 20px 235px; 
   		}

		#menu_content {
		    margin: 10px 5px 20px 16%; 
		    padding: 5px; 
   		}

		#right_text { 
			float: left; 
    		width: 100px; 
    	    margin: 70px 10px 20px 5px; 
		} 	
		
		#file { 	
			margin: 10px 5px 20px 235px; 
			padding: 5px;
		} 
	</style>
	<script src="lib/jquery-1.8.0.min.js"></script>
	<script type = "text/javascript">	
		load_information = null;
		var genomeCount = new Object();

		function viewTree(idform) { 
			if (document.getElementById(document.forms[idform].name + "_showTree_info").firstChild == null) { 
				load_information(document.forms[idform].elements["trees"].value, document.forms[idform].name);
			} 
		} 

		function loadFile(file) {
            var X = new XMLHttpRequest();
            X.open('GET', file, false );
            X.setRequestHeader('Content-Type', 'text/html')
            X.send('');
            return X.responseText;
        }
		
		function clearFile() { 
			document.getElementById('genome_file').value = '';
		} 
				 
		function setReconstructed(idform) { 
			document.forms[idform].elements["information_reconstracted"].value = document.forms[idform].reconstracted.checked;
		} 

		function showTarget(idform, show) {
			document.getElementById(document.forms[idform].name + "_targetDiv").style.display= show ? '' : 'none';
			document.getElementById(document.forms[idform].name + "_targetControl").style.display=show ? 'none' : '';
            document.forms[idform].elements["useTarget"].value=show ? 1 : 0;
			return false;
        }

		function addGenome(idform) {
			var form = document.forms[idform]; 
			++genomeCount[form.name];
			var value = genomeCount[form.name];
            var table = document.getElementById(form.name + "_genomes");
            var row = table.insertRow(value - 1);
            row.align = "left";
            row.vAlign = "top";
            var c = [];
            for (var j = 0; 1 > j; j++) {
                c[j] = row.insertCell(j);
                c[j].align = "left";
                c[j].vAlign = "top";
            }

			if(form.name == "grmForm") { 
       		    c[0].innerHTML = "<div id = \"right_text\"><strong>Genome " + value + ":</strong></div>" +
            	"<div id = \"content\"> <div>Name: <input size='1' id='name" + value + "' name='name" + value +
				"'> Aliases <input id='alias" + value + "'size='29' name='alias" + value + "'></div>" +
            	"<textarea id='genome" + value + "' name='genome" + value + "' rows='6' cols='61'></textarea></div>";
			} else { 
				 c[0].innerHTML = "<div id = \"content_text1\"><strong>Genome " + value + "</strong></div>" + 
				 "<div id = \"content\"> <div>Name: <input size='1' id='name" + value + "' name='name" + value +
	             "'> Aliases <input id='alias" + value + "'size='29' name='alias" + value + "'></div></div>"
			} 
		}	

		function initGenomes(idform, n) {
			var table = document.getElementById(document.forms[idform].name + "_genomes");
			for(var i = table.rows.length - 1; i >= 0; --i) {
                table.deleteRow(i);
            }           
			genomeCount[document.forms[idform].name] = 0;            
            for (var i = 0; i < n; i++) {
                addGenome(idform);
            }
        }
		
		function viewInformForExample(form, showStr) { 
			if (form.name == "infForm") { 
				document.getElementById("mam6_information").style.display = "mam6" == showStr ? "" : "none";
				document.getElementById("mam7_information").style.display = "mam7" == showStr ? "" : "none";
				document.getElementById("rosaceae_information").style.display = "rosaceae" == showStr ? "" : "none";
			} else { 
				document.getElementById("murphy_information").style.display = "murphy" == showStr ? "" : "none";
				document.getElementById("toy_information").style.display = "toy" == showStr ? "" : "none";
				document.getElementById("shigella_information").style.display = "shigella" == showStr ? "" : "none";
			}
		} 

		function clearFormat(idform) { 
			initGenomes(idform, 2);
			if (document.forms[idform].name == "infForm") { 
				document.forms[idform].elements["genome"].value = '';
			}
			document.forms[idform].elements["trees"].value = '';
            document.forms[idform].elements["stages"].value = '3';
            document.forms[idform].elements["target"].value = '';
            document.forms[idform].elements["completion"].value = '';
			document.forms[idform].reconstracted.checked = false;
			if (document.forms[idform].name == "fileForm") { 
				clearFile();
			} else { 
				viewInformForExample(document.forms[idform], "");
			} 			

			document.getElementById(document.forms[idform].name + "_showTree_info").innerHTML = "";
			showTarget(idform, false);
		} 

		function setFormat(format) { 
			for(var i = 1; i < document.forms.length; ++i) { 
				clearFormat(i);  				
			} 
			
			for(var i = 1; i < document.forms.length; ++i) { 
				if (document.forms[i].name == format) { 
					document.forms[i].style.display = ''; 
					document.choose.inputformat[i - 1].checked = true;
				} else { 
					document.forms[i].style.display = 'none'; 
					document.choose.inputformat[i - 1].checked = false;
				} 
			} 
			
			document.getElementById('showHelp').style.display = (format == 'help')?'':'none'; 
		} 

		/*Set example in file*/
	    function setMam6() {
            setFormat('infForm'); 
			initGenomes(2, 6); 
			var form = document.forms[2];
			viewInformForExample(form, "mam6");
			
			form.elements["name1"].value = 'M';
	        form.elements["name2"].value = 'R';
            form.elements["name3"].value = 'D';
            form.elements["name4"].value = 'Q';
            form.elements["name5"].value = 'H';
            form.elements["name6"].value = 'C';

            form.elements["alias1"].value = 'mouse mm9';
            form.elements["alias2"].value = 'rat rn4';
            form.elements["alias3"].value = 'dog canFam2';
            form.elements["alias4"].value = 'macaque rheMac2';
            form.elements["alias5"].value = 'human hg18';
            form.elements["alias6"].value = 'chimpanzee panTro2';

            form.elements["genome"].value = loadFile('mam6/blocks6.txt');
			form.elements["trees"].value = '(((M,R),D),(Q,(H,C)))';
			form.reconstracted.checked = false;
			setReconstructed(2);

            form.elements["stages"].value = '4';
            form.elements["target"].value = '';
            form.elements["completion"].value = '1299h 1300h 1300t 1301t R\n'+
                                                            '1289h 1290h 1290t 1291t D\n'+
                                                            '374t 376t oo 375h R\n'+
                                                            '927t 610h 610t 609h R\n'+
                                                            '79h 81t 80t 83h Q\n'+
                                                            '81t 83h 80h 84t Q\n'+
                                                            '794h 795t 795h 796t Q\n'+
                                                            '1196h 1198h 1202t 1199t D\n';
            showTarget(2, true);
        }

        function setMam7() {
            setFormat('infForm');
			initGenomes(2, 7); 
			var form = document.forms[2];
			viewInformForExample(form, "mam7");
			
			form.elements["name1"].value = 'M';
            form.elements["name2"].value = 'R';
            form.elements["name3"].value = 'D';
            form.elements["name4"].value = 'Q';
            form.elements["name5"].value = 'H';
            form.elements["name6"].value = 'C';
            form.elements["name7"].value = 'O';

            form.elements["alias1"].value = 'mouse mm9';
            form.elements["alias2"].value = 'rat rn4';
            form.elements["alias3"].value = 'dog canFam2';
            form.elements["alias4"].value = 'macaque rheMac2';
            form.elements["alias5"].value = 'human hg18';
            form.elements["alias6"].value = 'chimpanzee panTro2';
            form.elements["alias7"].value = 'opossum monDom4'; 
            
            form.elements["genome"].value = loadFile('mam7/blocks7.txt');
            form.elements["trees"].value = '((M,R),DOQHC)\n' +
                                           '(MRDO,(Q,(H,C)))';

			form.reconstracted.checked = true;
			setReconstructed(2);
			form.elements["stages"].value = '3';
            form.elements["target"].value = '';
            form.elements["completion"].value = '';

            showTarget(2, false);
        }

		function setRosaceae() { 
			setFormat('infForm'); 
			initGenomes(2, 3); 
			var form = document.forms[2];
			viewInformForExample(form, "rosaceae");
		
			form.elements["name1"].value = 'P';
            form.elements["name2"].value = 'S';
            form.elements["name3"].value = 'V';

            form.elements["alias1"].value = 'pp prunus';
            form.elements["alias2"].value = 'fv fragaria';
            form.elements["alias3"].value = 'vv vitus';
            
            form.elements["genome"].value = loadFile('rosaceae/rosaceae.txt');
            form.elements["trees"].value = '((P,S),V)';
			form.reconstracted.checked = false;
			setReconstructed(2);
            form.elements["stages"].value = '3';
            form.elements["target"].value = 'PS';
            form.elements["completion"].value = '';

            showTarget(2, true);
		} 


        function setMurphy() {
			setFormat('grmForm');
            initGenomes(3, 7);
			var form = document.forms[3];
			viewInformForExample(form, "murphy");
		
			form.elements["name1"].value = 'H';
         	form.elements["alias1"].value = 'human';
			form.elements["genome1"].value = loadFile('murphy05/murphy05_1.txt');

            form.elements["name2"].value = 'M';
            form.elements["alias2"].value = 'mouse';
			form.elements["genome2"].value = loadFile('murphy05/murphy05_2.txt');

			form.elements["name3"].value = 'R';
            form.elements["alias3"].value = 'rat';
			form.elements["genome3"].value = loadFile('murphy05/murphy05_3.txt');

            form.elements["name4"].value = 'C';
            form.elements["alias4"].value = 'cat';
			form.elements["genome4"].value = loadFile('murphy05/murphy05_4.txt');

            form.elements["name5"].value = 'W';
            form.elements["alias5"].value = 'cattle';
			form.elements["genome5"].value = loadFile('murphy05/murphy05_5.txt');

            form.elements["name6"].value = 'D';
            form.elements["alias6"].value = 'dog';
			form.elements["genome6"].value = loadFile('murphy05/murphy05_6.txt');

            form.elements["name7"].value = 'P';
            form.elements["alias7"].value = 'pig';
			form.elements["genome7"].value = loadFile('murphy05/murphy05_7.txt');           

            form.elements["trees"].value = '((H,(M,R)),DCPW)\n'+
                                           '((D,C),HRMPW)\n' +
                                           '((P,W),HRMDC)';
			form.reconstracted.checked = false;
			setReconstructed(3);
            form.elements["stages"].value = '3';
            form.elements["target"].value = 'DCPW';
            form.elements["completion"].value = '';
            showTarget(3, true);
        }

        function setToy() {
			setFormat('grmForm');
			initGenomes(3, 4);
			var form = document.forms[3];
			viewInformForExample(form, "toy");
		
     		form.elements["name1"].value = '1';
			form.elements["alias1"].value = 'P1';
            form.elements["genome1"].value = loadFile('toy/toyblks_1.txt');

			form.elements["name2"].value = '2';
            form.elements["alias2"].value = 'P2';
			form.elements["genome2"].value = loadFile('toy/toyblks_2.txt');

			form.elements["name3"].value = '3';
            form.elements["alias3"].value = 'P3';
			form.elements["genome3"].value = loadFile('toy/toyblks_3.txt');

            form.elements["name4"].value = '4';
            form.elements["alias4"].value = 'P4';
    		form.elements["genome4"].value = loadFile('toy/toyblks_4.txt');
			
            form.elements["trees"].value = '((1,2),(3,4))';
			form.reconstracted.checked = false; 
			setReconstructed(3);

            form.elements["stages"].value = '3';
            form.elements["target"].value = '';
            form.elements["completion"].value = '';

            showTarget(3, false);
        }

		function setShigella() { 
			setFormat('grmForm'); 
			initGenomes(3, 6);
			var form = document.forms[3];
			viewInformForExample(form, "shigella");
		
      		form.elements["name1"].value = 'A';
			form.elements["alias1"].value = 'Shi1 S1';
			form.elements["genome1"].value = loadFile('shigella/shigella_1.txt');

            form.elements["name2"].value = 'B';
            form.elements["alias2"].value = 'Shi2 S2';
			form.elements["genome2"].value = loadFile('shigella/shigella_2.txt');

			form.elements["name3"].value = 'C';
            form.elements["alias3"].value = 'Shi3 S3';
			form.elements["genome3"].value = loadFile('shigella/shigella_3.txt');

            form.elements["name4"].value = 'D';
            form.elements["alias4"].value = 'Shi4 S4';
			form.elements["genome4"].value = loadFile('shigella/shigella_4.txt');

            form.elements["name5"].value = 'E';
            form.elements["alias5"].value = 'Shi5 S5';
			form.elements["genome5"].value = loadFile('shigella/shigella_5.txt');

            form.elements["name6"].value = 'F';
            form.elements["alias6"].value = 'Shi6 S6';
			form.elements["genome6"].value = loadFile('shigella/shigella_6.txt');           

            form.elements["trees"].value = '(((A,B),F),((D,C),E))';
			form.reconstracted.checked = false;
			setReconstructed(3);
            
			form.elements["stages"].value = '4';       
            form.elements["completion"].value = '357h 365t 268t 321h D\n' + 
															  '400t 401h 398h 401t D\n' + 
															  '425t 426t 426h 407h B';
            showTarget(3, true);
		} 

		$(document).ready(function(){
			function my_load_func(trees, form) {    			
				$.ajax({ 
					type: "POST",
					url: "http://127.0.0.1:8080/showtree.html",
					async: false,
					cashe: false,
					context: document.body,
					data: "trees="+trees+"&width="+$("#"+form+"_showTree_info").width(),
					dataType: 'html',
					beforeSend: function() { 
						$("#" + form + "_showTree_bar").html("<u>Please wait. We processed this request: read information, generate images or html text, send. This may take some time.</u>");	
					}, 
					success: function(data) { 
						$("#" + form + "_showTree_info").html(data);
					}, 
					error: function() {
						$("#" + form + "_showTree_info").html("<p><strong>We can not create image. You can download information in view of text file</strong><p>");	 				
					}, 
					complete: function() { 
						$("#" + form + "_showTree_bar").html("");	
					}  	
				});
			} 
			load_information = my_load_func; 
		});
	</script>
</head>
<body>
	<header> 
		<h1><p align="center"><a href="#">MGRA (Multiple Genome Rearrangements and Ancestors) web server, beta version</a></p></h1>
	</header>
		
	<nav id = "menu">
		<p> 
			<strong> Input formats: </strong> <br>
			<blockquote> 
				<form name="choose">
					<input type="radio" name="inputformat" value="download file" onClick="setFormat('fileForm');" CHECKED/>  <font color = "blue"> download file </font> <br/>
					<input type="radio" name="inputformat" value="infercars" onClick="setFormat('infForm');"/>  <font color = "blue"> infercars </font> <br/>
					<input type="radio" name="inputformat" value="grimm" onClick="setFormat('grmForm');"/>  <font color = "blue"> grimm </font> <br/>					
				</form>
			</blockquote>
			<strong> Examples: </strong> <br>
			<blockquote> 						
				<a href="#" onclick="setRosaceae(); return false;">rosaceae</a> <br>      
				<a href="#" onclick="setMam7(); return false;">mam7</a> <br>	
				<a href="#" onclick="setMam6(); return false;">mam6</a> <br>
		        <a href="#" onclick="setMurphy(); return false;">murphy05</a> <br>
		        <a href="#" onclick="setToy(); return false;">toy</a> <br>
				<a href="#" onclick="setShigella(); return false;">shigella</a> <br>      
			</blockquote>
			<a href="#" onclick="setFormat('help'); return false;">How to work</a> <br>			
		</p>
	</nav>

	<div id = "menu_content">	
	<!-- file format to detected-->
	<form method="post" action="http://127.0.0.1:8080" enctype="multipart/form-data" name="fileForm">
		<h2><p align="center">Genomes</p></h2>	
	
		<input type="hidden" name="useFormat" id="useFormat" value="auto"/>
		<input type="hidden" name="information_reconstracted" id="information_reconstracted" value="false"/>
		<input type="hidden" name="useTarget" id="useTarget" value="0"/>

		<table id="fileForm_genomes">
	    </table>

	    <script>
			document.choose.inputformat[0].checked = true; 
	        initGenomes(1, 2);
	    </script>
		
		<div align = "center"> <input name="add_genome" type="button" value="add genome" onclick="addGenome(1); return false;"/> </div>

		<p>
        	<div id = "content_text1"> <strong>Load file with genomes:</strong> </div> 
			<div id = "file"> <input type="file" id="genome_file" name="genome_file"> <a href="#" onclick="clearFile();">Clear</a> </div>
	    </p>		

		<p>
    	    <strong>Known subtrees of the phylogenetic tree of the above genomes in <a href=http://en.wikipedia.org/wiki/Newick_format>Newick format</a></strong><br>
    	    MGRA assumes that only terminal branches are known.  
    	</p>
   		
		<div id = "content"> <textarea rows="5" cols="60" name="trees" id="trees"></textarea></div>
		<div align = "center"> <input name="show_tree" type="button" value="show tree" onclick="viewTree(1)"/> </div>
		<p id="fileForm_showTree_bar" align="center"></p>
		<div id="fileForm_showTree_info" align="center"></div>
	
	    <input type="checkbox" name="reconstracted" value="1" onclick="setReconstructed(1);"/><strong>&nbsp;recostructed tree</strong><br/>
		
	    <p>
	        <strong>The number of stages to perform:</strong>&nbsp;&nbsp;&nbsp;<input type='text' size='5' name='stages' id="stages" value="3"/><br>
	     	Higher values correspond to less reliable heuristics.
		    The highest value is 4 and usually it still requires manual completion (see below).
		    Smaller values require 'target' specification to produce ancestral reconstruction.
	      	It is not recommended to have stages higher than 3 if no complete phylogenetic tree is known.
		</p>
   		<p>	
			<strong>Completion</strong><br/>
		    Manual (human-assisted) completion of T-consistent transformation (the last stage of MGRA).
		    Each row contains exactly five terms: `a b c d E' denoting a 2-break operating on the \vec{T}-consistent multicolor E,
		    replacing the multiedges (a,b) and (c,d) with the multiedges (a,c) and (b,d).
		    'oo' stands for a chromosome end.
			Manual completion can be derived from visual inspection of the breakpoint graph after MGRA Stage 3.
		</p> 
		<div id = "content">
		    <textarea rows="5" cols="60" name="completion" id="completion"></textarea>
		</div>

	    <div id="fileForm_targetControl">
            <a href="#" onclick="return showTarget(1, true);">Show Target and Completion fields</a>
        </div>

        <div id="fileForm_targetDiv" style="display:none;">
    		<p>
    	    	<a href="#" onclick="return showTarget(1, false);">Remove Target and Completion fields</a>
	   		</p>

    		<p>
				<p>
					By default, MGRA reconstructs all ancestral genomes but it first has to complete transformation into an identity breakpoint graph.
		        	If 'target' is specified, MGRA focuses on reconstruction of the specified ancestral genome
		        	and can produce a partial reconstruction at any point (even if the transformation is not complete).
				</p>
    	    	<strong>Target:</strong>&nbsp;&nbsp;&nbsp;<input type='text' size='10' name='target' id="target"/><br>
    	   	</p>
		</div> 
			<p>
				<div align = "center"> 
			        <input type="submit" value="run mgra"/>
				</div>
		    </p>
    </form>
	
	<!--Form for infercars format-->
	<form method="post" action="http://127.0.0.1:8080" enctype="multipart/form-data" name="infForm" style="display:none;">
		
		<h2><p align="center">Genomes in infercars format</p></h2>	

		<input type="hidden" name="useFormat" id="useFormat" value="infercars"/>
		<input type="hidden" name="information_reconstracted" id="information_reconstracted" value="false"/>
	    <input type="hidden" name="useTarget" id="useTarget" value="0"/>
	
		<div id="mam6_information" style="display:none;">
			<p style="text-indent: 25px;"> <strong> Information on the example: </strong> </p>
			<p>
			Synteny blocks for 6 mammalian species, constructed by Jian Ma from the following builds: hg18 (Mar 2006), panTro2 (Mar 2006), rheMac2 (Jan 2006), mm9 (Jul 2007), rn4 (Nov 2004), canFam2 (May 2005). <br/> 
			Blocks 1268, 1353, and 1360 (representing intermixed segments of the chromosome X and other chromosomes) are deliberately commented out to simplify analysis of the chromosome X reconstruction. <br/>	
			<u>Paper:</u> Ma, J., Zhang, L., Suh, B. B., Raney, B. J., Burhans, R. C., Kent, J. W., Blanchette, M., Haussler, D., and Miller, W., 2006. Reconstructing contiguous regions of an ancestral genome. Genome Research, 16(12):1557–1565
			</p>
		</div>

		<div id="mam7_information" style="display:none;">
			<p style="text-indent: 25px;"> <strong> Information on the example: </strong> </p>
			<p>
			Synteny blocks for 7 mammalian species, constructed by Jian Ma from the following builds: hg18 (Mar 2006), panTro2 (Mar 2006), rheMac2 (Jan 2006), mm9 (Jul 2007), rn4 (Nov 2004), canFam2 (May 2005), monDom4 (Jan 2006). <br/>
			<u>Paper:</u> Ma, J., Zhang, L., Suh, B. B., Raney, B. J., Burhans, R. C., Kent, J. W., Blanchette, M., Haussler, D., and Miller, W., 2006. Reconstructing contiguous regions of an ancestral genome. Genome Research, 16(12):1557–1565
			</p>
		</div>

		<div id="rosaceae_information"	style="display:none;">
			<p style="text-indent: 25px;"> <strong> Information on the example: </strong> </p>
			<p>
				Synteny blocks for 3 genomes in rosaceous subfamilies: Prunus, Fragaria, Vitis. <br/>
				<u>Paper:</u> Sook Jung, Alessandro Cestaro, Michela Troggio, Dorrie Main, Ping Zheng, Ilhyung Cho, Kevin M Folta, Bryon Sosinski, Albert Abbott, Jean-Marc Celton, Pere Arús, Vladimir Shulaev, Ignazio Verde, Michele Morgante, Daniel Rokhsar, Riccardo Velasco and Daniel J Sargent, 4 April 2012. Whole genome comparisons of Fragaria, Prunus and Malus reveal different modes of evolution between Rosaceous subfamilies.
			</p>
		</div>
		
		<table id="infForm_genomes">
	    </table>

		<div align = "center"> <input name="add_genome" type="button" value="add genome" onclick="addGenome(2); return false;"/> </div>

		<div id ="right_text"> <strong>Genome:</strong></div>
		<div id ="content"> <textarea id='genome' name='genome' rows='10' cols='61'></textarea></div>

		<p>
    	    <strong>Known subtrees of the phylogenetic tree of the above genomes in <a href=http://en.wikipedia.org/wiki/Newick_format>Newick format</a></strong><br>
    	    MGRA assumes that only terminal branches are known.  
    	</p>
   		
		<div id = "content"> <textarea rows="5" cols="60" name="trees" id="trees"></textarea></div>
		<div align = "center"> <input name="show_tree" type="button" value="show tree" onclick="viewTree(2)"/> </div>
		<p id="infForm_showTree_bar" align="center"></p>
		<div id="infForm_showTree_info" align="center"></div>

		<input type="checkbox" name="reconstracted" value="1" onclick="setReconstructed(2);"/><strong>&nbsp;recostructed tree</strong><br/>
		
	    <p>
	        <strong>The number of stages to perform:</strong>&nbsp;&nbsp;&nbsp;<input type='text' size='5' name='stages' id="stages" value="3"/><br>
	        Higher values correspond to less reliable heuristics.
		    The highest value is 4 and usually it still requires manual completion (see below).
		    Smaller values require 'target' specification to produce ancestral reconstruction.
	        It is not recommended to have stages higher than 3 if no complete phylogenetic tree is known.
	    </p>
		
   		<p>	
			<strong>Completion</strong><br/>
	        Manual (human-assisted) completion of T-consistent transformation (the last stage of MGRA).
	        Each row contains exactly five terms: `a b c d E' denoting a 2-break operating on the \vec{T}-consistent multicolor E,
	        replacing the multiedges (a,b) and (c,d) with the multiedges (a,c) and (b,d).
	        'oo' stands for a chromosome end.
	        Manual completion can be derived from visual inspection of the breakpoint graph after MGRA Stage 3.
	    </p> 
		<div id = "content">
			<textarea rows="5" cols="60" name="completion" id="completion"></textarea>
		</div>	

	    <div id="infForm_targetControl">
            <a href="#" onclick="return showTarget(2, true);">Show Target and Completion fields</a>
        </div>

        <div id="infForm_targetDiv" style="display:none;">
    		<p>
    	    	<a href="#" onclick="return showTarget(2, false);">Remove Target and Completion fields</a>
	   		</p>
    		<p>
				<p>
					By default, MGRA reconstructs all ancestral genomes but it first has to complete transformation into an identity breakpoint graph.
		        	If 'target' is specified, MGRA focuses on reconstruction of the specified ancestral genome
		        	and can produce a partial reconstruction at any point (even if the transformation is not complete).
				</p>
    	    	<strong>Target:</strong>&nbsp;&nbsp;&nbsp;<input type='text' size='10' name='target' id="target"/><br>
    	   	</p>
		</div> 
			<p>
				<div align = "center"> 
			        <input type="submit" value="run mgra"/>
				</div>
		    </p>
    </form>

	<!--Form for grimm format-->
	<form method="post" action="http://127.0.0.1:8080" enctype="multipart/form-data" name="grmForm" style="display:none;">
		
		<h2><p align="center">Genomes in grimm format</p></h2>	

		<input type="hidden" name="useFormat" id="useFormat" value="grimm"/>
		<input type="hidden" name="information_reconstracted" id="information_reconstracted" value="false"/>
		<input type="hidden" name="useTarget" id="useTarget" value="0"/>
	
		<div id="murphy_information" style="display:none;">
			<p style="text-indent: 25px;"><strong>Information on the example:</strong></p>
			<p>
			Synteny blocks for 7 mammalian genomes:	human, rat, mouse, dog, cat, pig, and cow <br/>
			<u>Paper:</u> Murphy,W. J., Larkin, D. M., van derWind, A. E., Bourque, G., Tesler,G., Auvil, L., Beever, J. E., Chowdhary, B. P., Galibert, F., Gatzke, L., et al., 2005. Dynamics of Mammalian Chromosome Evolution Inferred from Multispecies Comparative Map. Science, 309(5734):613–617.
			</p>
		</div>	

		<div id="toy_information" style="display:none;">
			<p style="text-indent: 25px;"> <strong>Information on the example:</strong> </p>
			<p>
			Synteny blocks for a set of 4 toy genomes. <br/>
			<u>Paper:</u> Max Alekseyev, Pavel Pevzner "Breakpoint Graphs and Ancestral Genome Reconstructions." Genome Research, special issue "Genomics and Darwinism", 2009.
			</p>
		</div>	

		<div id="shigella_information" style="display:none;">
			<p style="text-indent: 25px;"> <strong>Information on the example:</strong> </p>
			<p>
			Synteny blocks constructed for the 6 strains Shigella spp (Shigella sonnei Ss046, Shigella boydii Sb227, Shigella boydii CDC 3083 94, Shigella flexneri 2a, Shigella flexneri 5 8401, Shigella flexneri 2a 2457T). <br/>
			<u>Authors:</u> Olga Bochkareva, Marat Kazanov, Mikhail Gelfand, personal communication. Institute for Information Transmission Problems of the Russian Academy of Sciences (Kharkevich Institute). 
			</p>
		</div>	

    	<table id="grmForm_genomes">
	    </table>
		
		<div align = "center"> <input name="add_genome" type="button" value="add genome" onclick="addGenome(3); return false;"/> </div>
	    
    	<p>
    	    <strong>Known subtrees of the phylogenetic tree of the above genomes in <a href=http://en.wikipedia.org/wiki/Newick_format>Newick format</a></strong><br>
    	    MGRA assumes that only terminal branches are known.  
    	</p>
   	
		<div id = "content"> <textarea rows="6" cols="60" name="trees" id="trees"></textarea></div>
		<div align = "center"> <input name="show_tree" type="button" value="show tree" onclick="viewTree(3)"/> </div>
		<p id="grmForm_showTree_bar" align="center"></p>
		<div id="grmForm_showTree_info" align="center"></div>

	    <input type="checkbox" name="reconstracted" value="1" onclick="setReconstructed(3);"/><strong>&nbsp;recostructed tree</strong><br/>
	    
	    <p>
	        <strong>The number of stages to perform:</strong>&nbsp;&nbsp;&nbsp;<input type='text' size='5' name='stages' id="stages" value="3"/><br>
	      	Higher values correspond to less reliable heuristics.
		    The highest value is 4 and usually it still requires manual completion (see below).
		    Smaller values require 'target' specification to produce ancestral reconstruction.
	      	It is not recommended to have stages higher than 3 if no complete phylogenetic tree is known.
	    </p>

 		<p>	
			<strong>Completion</strong><br/>
			Manual (human-assisted) completion of T-consistent transformation (the last stage of MGRA).
		    Each row contains exactly five terms: `a b c d E' denoting a 2-break operating on the \vec{T}-consistent multicolor E,
		    replacing the multiedges (a,b) and (c,d) with the multiedges (a,c) and (b,d).
		    'oo' stands for a chromosome end.
		    Manual completion can be derived from visual inspection of the breakpoint graph after MGRA Stage 3.
		</p>
		<div id = "content">
		    <textarea rows="6" cols="60" name="completion" id="completion"></textarea>
		</div>

	    <div id="grmForm_targetControl">
            <a href="#" onclick="return showTarget(3, true);">Show Target and Completion fields</a>
        </div>

        <div id="grmForm_targetDiv" style="display:none;">
    		<p>
    	    	<a href="#" onclick="return showTarget(3, false);">Remove Target and Completion fields</a>
	   		</p>
    		<p>
				<p>
					By default, MGRA reconstructs all ancestral genomes but it first has to complete transformation into an identity breakpoint graph.
		        	If 'target' is specified, MGRA focuses on reconstruction of the specified ancestral genome
		        	and can produce a partial reconstruction at any point (even if the transformation is not complete).
				</p>
    	    	<strong>Target:</strong>&nbsp;&nbsp;&nbsp;<input type='text' size='10' name='target' id="target"/><br>
    	   	</p>
		</div>
			<p>
				<div align = "center"> 
			        <input type="submit" value="run mgra"/>
				</div>
		    </p>
    </form>
	</div>

	<div id = "showHelp" style="display:none;"> <div id = "content">
		<h2><p align="center">How to work</p></h2>		
    	This is web server provide interface for working <a href="http://cseweb.ucsd.edu/~ppevzner/download/mgra20090215.zip">MGRA algorithm</a>. 
		<h3><u>Algorithm:</u></h3>		
		<ol type = "1">
		<li>
			<strong>You need choose input data format. Server supported two input formats are grimm and infercars. </strong> 
		</li>
		<br>
		<li>
			<strong>You enter their data in form:</strong> 
			<ul type="square">
			<li><u>Grimm form</u> <br>
				Name of genome must be a single unique letter.
			</li>
			<br>
			<li><u>Infercars form</u> <br>
				Name of genome must be a single unique letter.
			</li>
			</ul>
			<br>
			We recomended see examples (mam6, mam7, rosaceae - infercars format, murphy and toy - grimm format) before they enter their data. 	
		</li>
		<br>
		<li>
			<strong> Push button "run mgra", and server will begin processing the data, which may take some time. </strong>
		</li>
		<br>
		<li>
			<strong> Information on what step is processing the data will be displayed on the screen. </strong> 
		</li>
		<br>
		<li>
			<strong> After completion of data processing, it will display the phylogenetic tree. </strong> <br>
			Tree nodes can be placed as you wish by dragging the mouse. <br> 
			If you double click on the yellow tree node, you saw information of genome and genome transformation.    
			<ul type="square">
			<br>
			<li><u>Genome</u> <br>
				Genome represented as the sequence of synteny blocks. Each synteny block has its number and oriented to the head or to the tail. If we will use numbers for the representation it will be pretty hard to figure the length of the genome or its part. That's why we use > and &lt; to represent synteny blocks, depending on their orientation. The synteny block number can be viewed as title to the link associated with > or 					&lt; sign. 
			</li>
			<br>
			<li><u>Genome transformation</u> <br>
				The original MGRA outputs only the ends of the breaks of the transformations. We show this ends using h and t letters. Each end has its own color that allows us to show where it goes after the transformation. We use similar colors for the ends that are clued after the transformation.
			</li>
		</li>
		<ol>

		Name genome is one symbol
	</div>
	</div>

	<footer>
	<hr>
	MGRA 1.0 &copy; 2008,09 by Max Alekseyev
	</footer>
</body>
</html>

